File: Clean-Architecture/Feature/Javascript/example/src/features/auth/data/repositories/AuthRepositoryImpl.js
import { User } from '../../domain/entities/User.js';
import { AuthRepository } from '../../domain/repositories/AuthRepository.js';

export class AuthRepositoryImpl extends AuthRepository {
  constructor({ authDataSource }) {
    super();
    this.authDataSource = authDataSource;
  }

  async login(credentials) {
    const record = await this.authDataSource.findByCredentials(credentials);
    if (!record) return null;

    return {
      user: new User(record),
      token: `demo-token-${record.id}`,
    };
  }
}

File: Clean-Architecture/Feature/dart/example/lib/features/auth/data/repositories/auth_repository_impl.dart
import '../../../../core/error/failure.dart';
import '../../../../core/result/result.dart';
import '../../domain/entities/auth_session.dart';
import '../../domain/repositories/auth_repository.dart';
import '../datasources/auth_remote_data_source.dart';

class AuthRepositoryImpl implements AuthRepository {
  const AuthRepositoryImpl(this._remoteDataSource);
  final AuthRemoteDataSource _remoteDataSource;

  @override
  Future<Result<AuthSession>> login({required String email, required String password}) async {
    try {
      final result = await _remoteDataSource.login(email: email, password: password);
      return Success(AuthSession(user: result.user.toEntity(), token: result.token));
    } on ApiException catch (error) {
      return FailureResult(Failure(error.message, code: error.code));
    } catch (_) {
      return const FailureResult(Failure('Unable to connect to the server'));
    }
  }
}
